### PR TITLE
[DRAFT] Log gradient norms

### LINKED ISSUE #688: Log gradient norms to understand and debug model better
### Is your feature request related to a problem? Please describe.

See title, ideally with some plotting code, and potentially behind a flag to prevent large metric files.

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### Organisation

_No response_

### PR DESCRIPTION
**Pull request by sophie-xhonneux**
## Description

Add plotting and recording for grad norms

## Type of Change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update

## Issue Number

Closes #688 

## Code Compatibility

-   [ ] I have performed a self-review of my code

### Code Performance and Testing

-   [ ] I ran the `uv run train` and (if necessary) `uv run evaluate` on a least one GPU node and it works 
-   [ ] If the new feature introduces modifications at the config level, I have made sure to have notified the other software developers through Mattermost and updated the paths in the `$WEATHER_GENERATOR_PRIVATE` directory

### Dependencies

-   [ ] I have ensured that the code is still pip-installable after the changes and runs
-   [ ] I have tested that new dependencies themselves are pip-installable.
-   [ ] I have not introduced new dependencies in the inference portion of the pipeline

### Documentation

-   [ ] My code follows the style guidelines of this project
-   [ ] I have updated the documentation and docstrings to reflect the changes
-   [ ] I have added comments to my code, particularly in hard-to-understand areas

## Additional Notes
image
image
image

### CODE REVIEW

File: src/weathergen/train/trainer.py [hunk 1]
```diff
@@ -482,7 +483,19 @@ def train(self, epoch):
 
             # gradient clipping
             self.grad_scaler.unscale_(self.optimizer)
-            torch.nn.utils.clip_grad_norm_(self.ddp_model.parameters(), max_norm=cf.grad_clip)
+            total_norm = torch.nn.utils.clip_grad_norm_(
+                self.ddp_model.parameters(), max_norm=cf.grad_clip
+            )
+
+            # log gradient norms
+            if bidx % log_interval == 0:
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: Can we put this into a separate function `_log_grad_norms()`

File: src/weathergen/train/trainer.py [hunk 2]
```diff
@@ -76,6 +76,7 @@ def init(
 
         self.init_perf_monitoring()
         self.train_logger = TrainLogger(cf, config.get_path_run(self.cf))
+        self.last_grad_norm = 0.0
```
Comment (clessig) [inline src/weathergen/train/trainer.py]: Other quantities that we log are typically averaged, see [here](https://github.com/ecmwf/WeatherGenerator/blob/353401faae2cb9e9b3fa723afbfbf1b302bd5bd2/src/weathergen/train/trainer.py#L635). 

Here we want instantaneous values? But it would be nice to integrate it a bit better with the existing code in trainer, e.g. init self.last_grad_norm [here](https://github.com/ecmwf/WeatherGenerator/blob/353401faae2cb9e9b3fa723afbfbf1b302bd5bd2/src/weathergen/train/trainer.py#L460) if possible.

**Review by clessig — COMMENTED** [review]
I am happy to have this merged quickly since it is probably useful and quite orthogonal to other things. 

What is open for me is if we want to log averaged or instantaneous values. Also: there's no need to do multi-node averaging (as it is done [here](https://github.com/ecmwf/WeatherGenerator/blob/353401faae2cb9e9b3fa723afbfbf1b302bd5bd2/src/weathergen/train/trainer.py#L654)) since the gradient is averaged by DDP?

File: src/weathergen/train/trainer.py (hunk 1 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/trainer.py]: done

File: src/weathergen/train/trainer.py (hunk 2 above)
Comment (sophie-xhonneux) [reply src/weathergen/train/trainer.py]: Done

File: src/weathergen/utils/plot_grad_norms.py [hunk 3]
```diff
@@ -17,193 +19,202 @@ def __init__(self, json_file_path):
...
         self.load_data()
-        
+
     def load_data(self):
         """Load and parse the JSON data from file."""
         print(f"Loading data from {self.json_file_path}...")
-        
-        with open(self.json_file_path, 'r') as f:
+
+        with open(self.json_file_path) as f:
```
Comment (clessig) [inline src/weathergen/utils/plot_grad_norms.py:27]: Why has the `'r'` been removed. Read-only should be fine, not?

File: src/weathergen/train/trainer.py [hunk 4]
```diff
@@ -482,7 +485,13 @@ def train(self, epoch):
 
             # gradient clipping
             self.grad_scaler.unscale_(self.optimizer)
-            torch.nn.utils.clip_grad_norm_(self.ddp_model.parameters(), max_norm=cf.grad_clip)
+            total_norm = torch.nn.utils.clip_grad_norm_(
+                self.ddp_model.parameters(), max_norm=cf.grad_clip
+            )
+
+            # log gradient norms
+            if bidx % log_interval == 0 and self.log_grad_norms:
```
Comment (clessig) [inline src/weathergen/train/trainer.py:493]: Would be better to have a separate parameter log_interval_grads? Since it's instantaneous one might want to have it at 1 when one looks into a problem?

**Review by clessig — APPROVED** [review]
Left some minor comments. Ready to be merged after the DDP test, and after #696
